squint.c

```diff
@@ -0,0 +1,1611 @@
+         }
+         else { // str[b] rd, [fp, #X] -> mov rX, rd
+            *scan = 0xe1a00000 | ((BR+i) << 12) | ((*scan >> 12) & 0xf);
+         }
+      }
+   }
+
+   /* load frame vars into registers at top of function */
+   for (scan = funcBegin; *scan != NOP; ++scan);
+   for (i = 0; i < numReg; ++i) {
+      *scan++ = 0xe51b0000 | ((BR + i) << 12) |
+                ((offset[i] < 0) ? -offset[i] : (offset[i] | (1<<23)));
+   }
+}
+
+/**********************************************************/
+/********* Peephole optimization driver function **********/
+/**********************************************************/
+
+void peephole_opt(int *begin, int *end)
+{
+   int *scan = begin;
+   int *tmpbuf = (int *) malloc((end-begin)*sizeof(int));
+
+   create_const_map(begin, end);
+   const_imm_opt(begin, end);
+
+   while (scan < end) {
+      if (*scan == 0xe92d4800 && !is_const(scan)) { // push {fp, lr}
+         int *funcBegin = scan;
+         int *funcEnd;
+         int numReturn = 0;
+         int *retAddr = 0;
+         ++scan;
+         while (scan < end) {
+            if (*scan == 0xe92d4800 && !is_const(scan)) { // push {fp, lr}
+               break;
+            }
+            else if (*scan == 0xe8bd8800 && !is_const(scan)) { // pop {fp, pc}
+               ++numReturn;
+               retAddr = scan;
+            }
+            ++scan;
+         }
+         --scan;
+         funcEnd = scan /* retAddr */;
+
+         /******************************************/
+         /***   convert stack VM to frame VM     ***/
+         /******************************************/
+
+         simplify_branch(funcBegin, retAddr);
+         apply_peepholes1(funcBegin, retAddr);
+         apply_peepholes2(tmpbuf, funcBegin, retAddr);
+         apply_peepholes1(funcBegin, retAddr);
+         apply_peepholes3(tmpbuf, funcBegin, retAddr);
+         create_pushpop_map(tmpbuf, funcBegin, retAddr);
+         apply_peepholes4(funcBegin, retAddr);
+
+         /******************************************/
+         /***  convert frame VM to register VM   ***/
+         /******************************************/
+
+         rename_register1(funcBegin, retAddr);
+
+         rename_nop(funcBegin, retAddr);
+         funcEnd = relocate_nop(funcBegin, funcEnd, 0);
+      }
+      else {
+         ++scan;
+      }
+   }
+
+   destroy_const_map();
+   free(tmpbuf);
+}
+
+/**********************************************************/
+/******** read exe, optimize, write exe *******************/
+/**********************************************************/
+
+int main(int argc, char *argv[])
+{
+   int fd;
+   int offset, length;
+   int *mem;
+
+   if (argc != 4) {
+      printf("Use: %s <amacc executable> <text start> <length>\n", argv[0]);
+      exit(-1);
+   }
+
+   offset = strtol(argv[2], 0, 16);
+   length = strtol(argv[3], 0, 16);
+
+   if (!(offset > 0 && length > 0)) {
+       printf("last two arguments must be hex file offset and length\n");
```

Comment:
I see most lines in this file use 3-space indent, but this line has 7 space characters.

How about re-indent this file with 4 space character to follow amacc.c?

Reply:
The logic fits in 80 columns now, and I believe it is easier to comprehend this way than if I were to break it it up to reduce indenting,  The 3 space requirement is needed to stay in 80 columns. My goal was to improve readability/comprehension first, and performance second.

PS I noticed that the compile time was a feature listed in README.md, so I am also trying to keep runtime low for Squint.  I can improve performance more if needed.  Furthermore, Squint can make the executable size the same as the gcc compiler if I compress out nops between functions, and I can do that in this update if you would like.  The relocate_nop function was already designed to do that, I just need to complete it by fixing a bug.

Reply:
The default option of AMaCC enables fast code generation while squint allows generating faster code. It sounds pretty well to me. Let's finalize the current codebase.